compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1949,7 +1987,14 @@ class Typer extends Namer
         ctx.warning("Canceled quote directly inside a splice. ${ '{ XYZ } } is equivalent to XYZ.", tree.sourcePos)
         typed(innerExpr, pt)
       case expr =>
-        typedApply(untpd.Apply(untpd.ref(defn.InternalQuoted_exprSpliceR), tree.expr), pt)(spliceContext).withSpan(tree.span)
+        if (ctx.mode.is(Mode.QuotedPattern)) {
```

Comment:
We should only do this transformation at `level == 1` (i.e. the contents of the splice are at level 0)

Reply:
Agreed.